### PR TITLE
Sky-averaged Upper Limits

### PR DESCRIPTION
**Pull request by samueleronchini**

### CODE REVIEW

**Comment by Tohuvavohu** [comment]
@samueleronchini  this PR does not pass the checks, your changes break the install (click 'details' next to the failed tests below), you can also run these tests locally to see the issue.
Please fix.

Please better  separate the upper limit calculations, from the plotting.

also, in the future please provide more info in your commit messages and notes in the pull request documenting the changes.

**Comment by samueleronchini** [comment]
The checks fail because I use gbm tools to handle fermi maps. And that module is not installed. What if we add that in the requirements.txt?

When you say "Please better separate the upper limit calculations, from the plotting." you mean to create another code that does only plotting?
Because the plotting is already a separate function inside this code.

File: nitrates/analysis_seeds/UpperLimits.py [hunk 1]
```diff
@@ -1,28 +1,56 @@
...
 from ..lib.dbread_funcs import get_info_tab
 from ..lib.coord_conv_funcs import  convert_theta_phi2radec
 
-from ..analysis_seeds.do_full_rates import *
+from .do_full_rates import *
 
 from ..lib.calc_BAT_ul import *
 from ..config import resp_dname
 
+from gbm.data import HealPix
```
Comment (Tohuvavohu) [inline nitrates/analysis_seeds/UpperLimits.py:29]: why do you need these gbm healpix?

File: nitrates/analysis_seeds/UpperLimits.py [hunk 2]
```diff
@@ -35,21 +63,604 @@ def cli():
...
+        filename=os.path.join(path_results,'upper_limits_analysis.log'),
         level=logging.DEBUG,
         format="%(asctime)s-" "%(levelname)s- %(message)s",
     )
 
-    work_dir=args.work_dir
+
     #resp_dname=args.resp_dname
 
     conn = get_conn(os.path.join(work_dir,'results.db'))
```
Comment (Tohuvavohu) [inline nitrates/analysis_seeds/UpperLimits.py:666]: is this used?

File: nitrates/analysis_seeds/UpperLimits.py [hunk 3]
```diff
@@ -1,28 +1,56 @@
...
+import astropy_healpix as ah
+import astropy.units as u
+import healpy as hp
+import sys
+from ligo.gracedb.rest import GraceDb
+sys.path.append('/gpfs/group/jak51/default/UtilityBelt/UtilityBelt')
+from skyplot import get_earth_sat_pos
+import scipy.interpolate
+from scipy.spatial import Delaunay
+import pandas as pd
```
Comment (Tohuvavohu) [inline nitrates/analysis_seeds/UpperLimits.py:45]: you already have this

File: nitrates/analysis_seeds/UpperLimits.py [hunk 4]
```diff
@@ -1,28 +1,56 @@
...
 import argparse
 import time
+import numpy
+from datetime import datetime, timedelta
+from astropy.time import Time
+import json
+
 
 from ..lib.event2dpi_funcs import det2dpis, mask_detxy
 from ..lib.sqlite_funcs import get_conn
```
Comment (Tohuvavohu) [inline nitrates/analysis_seeds/UpperLimits.py:20]: do you need this?

File: nitrates/analysis_seeds/UpperLimits.py [hunk 5]
```diff
@@ -1,28 +1,56 @@
...
 from ..lib.dbread_funcs import get_info_tab
```
Comment (Tohuvavohu) [inline nitrates/analysis_seeds/UpperLimits.py:21]: do you need this?

File: nitrates/analysis_seeds/UpperLimits.py [hunk 6]
```diff
@@ -1,28 +1,56 @@
...
+from gbm.data import GbmHealPix
+
+from matplotlib.colors import LogNorm
+import ligo.skymap.plot
+import ligo.skymap.io
+from astropy.table import Table
+import astropy_healpix as ah
+import astropy.units as u
+import healpy as hp
+import sys
```
Comment (Tohuvavohu) [inline nitrates/analysis_seeds/UpperLimits.py:39]: you already have this above

**Review by Tohuvavohu — CHANGES_REQUESTED** [review]
@samueleronchini also i recommend moving the plotting component and a function to compute the upper limits into UtilityBelt instead,

File: nitrates/analysis_seeds/UpperLimits.py (hunk 2 above)
Comment (Tohuvavohu) [inline nitrates/analysis_seeds/UpperLimits.py:666]: do you need his?

File: nitrates/analysis_seeds/UpperLimits.py (hunk 3 above)
Comment (Tohuvavohu) [reply nitrates/analysis_seeds/UpperLimits.py:45]: pandas i mean

File: nitrates/analysis_seeds/UpperLimits.py (hunk 6 above)
Comment (Tohuvavohu) [reply nitrates/analysis_seeds/UpperLimits.py:39]: sys i mean

File: nitrates/analysis_seeds/UpperLimits.py (hunk 2 above)
Comment (Tohuvavohu) [reply nitrates/analysis_seeds/UpperLimits.py:666]: gihub is being supid. im referring to line 666, get_conn

**Comment by Tohuvavohu** [comment]
@samueleronchini why are you using gbmtools?

yes, seperate code. see also comment about moving things into UtilityBelt

File: nitrates/analysis_seeds/UpperLimits.py (hunk 1 above)
Comment (samueleronchini) [reply nitrates/analysis_seeds/UpperLimits.py:29]: that is used to create a map from a polygon, if given as input

Comment (Tohuvavohu) [reply nitrates/analysis_seeds/UpperLimits.py:29]: ok move all this stuff to UtilityBelt

File: nitrates/analysis_seeds/UpperLimits.py (hunk 2 above)
Comment (samueleronchini) [reply nitrates/analysis_seeds/UpperLimits.py:666]: that is the part of Gayathri, I think it's needed to get the trigger time

File: nitrates/analysis_seeds/UpperLimits.py (hunk 6 above)
Comment (samueleronchini) [reply nitrates/analysis_seeds/UpperLimits.py:39]: Right, I remove

File: nitrates/analysis_seeds/UpperLimits.py (hunk 4 above)
Comment (samueleronchini) [reply nitrates/analysis_seeds/UpperLimits.py:20]: Yes, I need all three

File: nitrates/analysis_seeds/UpperLimits.py (hunk 3 above)
Comment (samueleronchini) [reply nitrates/analysis_seeds/UpperLimits.py:45]: Right, thanks!

**Comment by samueleronchini** [comment]
Ok, then I will move everything to UtilityBelt, where I will split this in two codes, one for computation of upper limits and the other only for plotting maps. Right?

File: nitrates/analysis_seeds/UpperLimits.py (hunk 2 above)
Comment (Tohuvavohu) [reply nitrates/analysis_seeds/UpperLimits.py:664]: its not actually called again...

Comment (samueleronchini) [reply nitrates/analysis_seeds/UpperLimits.py:664]: conn is called here
https://github.com/Swift-BAT/NITRATES/blob/a48a8b74c511ea3e4a04dc7422f4db0942932dfb/nitrates/analysis_seeds/UpperLimits.py#L664C1-L668C35

**Comment by Tohuvavohu** [comment]
You cant move *everything* because you import things from other parts of NITRATES here...

**Comment by samueleronchini** [comment]
Can I move to UtilityBelt only the stuff that uses gbm tools and makes the plots?

**Comment by samueleronchini** [comment]
@Tohuvavohu  The checks give as error:
ModuleNotFoundError: No module named 'ul_skyaverage'
because I am trying to import from UtilityBelt, but ul_skyaverage is only in my branch. Is there a way to import correctly from UtilityBelt the module ul_skyaverage?

**Pull request closed without merging**